atakama/rule_engine.py

```diff
@@ -0,0 +1,195 @@
+"""Atakama keyserver ruleset library"""
+import abc
+from dataclasses import dataclass
+from enum import Enum
+from typing import List, Dict, Union, TYPE_CHECKING
+import logging
+
+import yaml
+
+from atakama import Plugin
+
+if TYPE_CHECKING:
+    from pathlib import Path
+
+log = logging.getLogger(__name__)
+
+
+class RequestType(Enum):
+    DECRYPT = "decrypt"
+    SEARCH = "search"
+    CREATE_PROFILE = "create_profile"
+    ACTIVATE_LOCATION = "activate_location"
+    CREATE_LOCATION = "create_location"
+    RENAME_FILE = "rename_file"
+    SECURE_EXPORT = "secure_export"
+
+
+@dataclass
+class ProfileInfo:
+    profile_id: bytes
+    """Requesting profile uuid"""
+    profile_words: List[str]
+    """Requesting profile 'words' mnemonic"""
+
+
+@dataclass
+class MetaInfo:
+    meta: str
+    """Typically the mount point sub-path of a file."""
+    complete: bool
+    """Whether the meta is complete (fully verified) or partial (missing components)"""
+
+
+@dataclass
+class ApprovalRequest:
+    request_type: RequestType
+    """Request type"""
+    device_id: bytes
+    """Requesting device uuid"""
+    profile: ProfileInfo
+    """"""
+    auth_meta: List[MetaInfo]
+    """Authenticated metadata associated with the encrypted data.   Typically a path to a file."""
+
+
+class RulePlugin(Plugin):
+    """
+    Base class for key server approval rule handlers.
+
+    When a key server receives a request, rules are consulted for approval.
+
+    Each rule receives its configuration from the policy file,
+    not the atakama config, like other plugins.
+    """
+
+    @abc.abstractmethod
+    def approve_request(self, request: ApprovalRequest) -> bool:
+        """
+        Return True if the request to decrypt a file will be authorized.
+
+        Return False if the request is to be denied.
+        Raise None if the request type is unknown or invalid.
+        Exceptions and None are logged, and considered False.
+
+        This is called any time:
+            a decryption agent wishes to decrypt a file.
+            a decryption agent wishes to search a file.
+            a decryption agent wishes to create a new multifactor profile.
```

Comment:
Per discussion this will be handled in default code elsewhere

Reply:
changed doc to refer to the RequestType enum